src/Sarif/EnumeratedArtifact.cs

```diff
@@ -16,44 +16,162 @@ public EnumeratedArtifact(IFileSystem fileSystem)
             FileSystem = fileSystem;
         }
 
+        internal byte[] bytes;
         internal string contents;
 
+        private bool? isBinary;
+        private Encoding encoding;
+
         public Uri Uri { get; set; }
 
+        public bool IsBinary
+        {
+            get
+            {
+                return this.isBinary ?? IsArtifactBinary();
+            }
+        }
+
+        internal bool IsArtifactBinary()
+        {
+            if (isBinary.HasValue)
+            {
+                return isBinary.Value;
+            }
+
+            GetArtifactData();
+            this.isBinary = this.contents == null;
+            return this.isBinary.Value;
+        }
+
         public Stream Stream { get; set; }
 
-        public Encoding Encoding { get; set; }
+        public Encoding Encoding
+        {
+            get
+            {
+                if (encoding == null)
+                {
+                    GetArtifactData();
+                }
+                return this.encoding;
+            }
+
+            set => this.encoding = value;
+        }
 
         internal IFileSystem FileSystem { get; set; }
 
+        /// <summary>
+        /// Gets or sets a property that determines whether the enumerated artifact
+        /// can obtain data from a non-seekable stream. This operation will be
+        /// less performant in any case where an artifact is a textual file (as
+        /// the file will first be converted to a byte array to determine
+        /// whether it is textual and subsequently turned into a string if it is).
+        /// </summary>
+        public bool SupportNonSeekableStreams { get; set; }
+
         public string Contents
         {
-            get => GetContents();
+            get => GetArtifactData().text;
             set => this.contents = value;
         }
 
-        private string GetContents()
+        public byte[] Bytes
+        {
+            get => GetArtifactData().bytes;
+            set => this.bytes = value;
+        }
+
+        private (string text, byte[] bytes) GetArtifactData()
         {
-            if (this.contents != null) { return this.contents; }
+            if (this.contents != null)
+            {
+                return (this.contents, bytes: null);
+            }
+
+            if (this.bytes != null)
+            {
+                return (text: null, this.bytes);
+            }
 
-            if (Stream == null && this.contents == null)
+            if (Stream == null && this.contents == null && this.bytes == null)
```

Comment:
>Stream [](http://example.com/codeflow?start=16&length=6)

nit: have both `Stream` and `this.Stream`. could be consistent by using 1 style `this.Stream`

Reply:
made this consistent for this file. We should discuss this style point, whether properties have a `this.` prefix in code is very inconsistent across the code base.